### DISCUSSION TITLE
Gasket type elements

### DISCUSSION THREAD

**Discussion by srinath-chakravarthy** [idaholab/moose#22082] [Q&A General] [answered]
Hi all, 
I would like to develop a gasket type element. One example of this is a QUAD4 element with 4 nodes, however with integration points lying on the mid-plane between bottom and top of the element. How would i go about doing this ?.I have looked at libmesh/elem and libmesh/gaussian_quad. So would i just define a new element QUAD4G and attach a default new quadrature rule to this element ? If so how would i integrate this into moose. 

 In essence this will be like a finite-thickness traditional cohesive element. The nice thing is the finite thckness version is non-singular and can presumably use all existing moose methods. 
I understand that there is a DG version of cohesive element in moose, but i am not able to decipher if this can be used for a finite thickness context. 

Cheers
Srinath

**Comment by srinath-chakravarthy** [comment, marked as answer]
Any suggestions anyone ?

**Reply by GiudGiud** [reply]
@roystgnr

**Reply by roystgnr** [reply]
There seems to be a bit of language difference here - the midplane between the top and bottom of a `Quad4` with planar nodes **is** the `Quad4` - it's a `Face`, a 2-D manifold (albeit not planar in the general case; you can have a "twist" between opposing edges) element. You can use it to solve an approximation to a 3-D problem, but that happens in the problem formulation (in the kernel), not the `Elem`.

Maybe that's a bigger language difference?  The combination of geometric `Elem` and `FE` space and physics kernel all together is what other idioms sometimes lump together as an "element" ... but IMHO trying to explicitly add every combination ahead of time is like insisting on finding the "element name" key on a keyboard that already has e,l,m,n,t, and a.

When I solve [shallow water flow](https://en.wikipedia.org/wiki/Shallow_water_equations) on a quad mesh, I don't need to define `QUAD4S` as a new `ElemType`; the depth-integrated equations went from 3D to 2D, so I just derive a weak form, add `FE` variables (which is what defines the "Finite Element" in the mathematical sense; it's not just the geometric element), and then I'd probably use Gaussian quadrature to solve those, but that all works with standard quads and/or triangles on a 2D domain, because while there's a 3D physics, in that formulation it's entirely defined by the 2D depth-averaged equations.  Or suppose I had a beam equation - I might add displacement variables with `FEType` of `HERMITE` to solve the $C^1$ weak form, or twice as many variables (e.g. displacements plus rotations) with some $C^0$ type like `LAGRANGE` to solve a mixed form, but either way I'd be solving on an ordinary `EDGE2` or `EDGE3` mesh.

In your case, then?  What you need to write depends on the shape function and quadrature details.  If you can describe a solution solely on an hex element's midplane and there's a canonical way to extrapolate from that quad midplane to the full hex, then you can almost certainly run on a standard `Quad4` or `Quad9`.  Can you write the space of shape functions on that plane with (a tensor product of!) existing spaces in libMesh?  If so then you're good to go there too; if not then that's what you'll end up having to write a new `FE` subclass and `FEType` for.  Are your quadrature points described by one of the standard rules?  Even if not then you definitely don't want to try to edit the Gaussian quadrature, which already has a standard definition on quads, but in that case you'll have to add your own `QBase` subclass and corresponding `QuadratureType`.

**Reply by hugary1995** [reply]
I believe there's a misunderstanding. @srinath-chakravarthy was asking for a genuine new type of quadrature rule. He was talking about `Quad4` in 2-D as a volume element instead of `Quad4` in 3-D as a face element. Instead of the standard Gaussian quadrature, he wants to place quadrature points along the midline (not midplane) of the `Quad4` element. So I think @roystgnr 's suggestion towards the end should apply.

What I have trouble understanding is that a `Quad4` element has two orthogonal midlines, how do you want to tell the new quadrature rule to use one over the other @srinath-chakravarthy ?

**Reply by hugary1995** [reply]
ok, I can see that Abaqus has its own convention :-)
https://classes.engineering.wustl.edu/2009/spring/mase5513/abaqus/docs/v6.6/books/usb/default.htm?startat=pt06ch26s06alm47.html

**Reply by roystgnr** [reply]
Oh, the mid*line*!  I was imaging a new type of shell element.  Thanks for clarifying!

Yeah, "your own `QBase` subclass and corresponding `QuadratureType`" would be the way to go here.

> a Quad4 element has two orthogonal midlines, how do you want to tell the new quadrature rule to use one over the other

That's actually doable in the mesh generator - since we always `map elem->point(0)` in physical space back to (-1,-1) in master space (and then go counter-clockwise from there), a quadrature rule that always returns points on the xi axis will always return points on the midline bisecting edges 03 and 12; a rule that always returns points on the eta axis will always return points on the line bisecting 01 and 32

**Reply by srinath-chakravarthy** [reply]
@hugary1995  thanks for clarifying and @roystgnr for the suggestions. The quadrature point definition is rather easy on the libmesh side. How would i go about the moose part of it, namely attaching this quadrature rule to the Quad4 element ? 
> That's actually doable in the mesh generator - since we always map elem->point(0) in physical space back to (-1,-1) in master space (and then go counter-clockwise from there), a quadrature rule that always returns points on the xi axis will always return points on the midline bisecting edges 03 and 12; a rule that always returns points on the eta axis will always return points on the line bisecting 01 and 32", 
This would mean i have to make sure that the physical point(0) and side (0,3) and side (1,2) is oriented in the direction of interest ?

**Reply by hugary1995** [reply]
I think once you set up your quadrature rule, you just need to add it to `enum_quadrature_type.h`. Then the `SetupQuadratureAction` in MOOSE should be able to pick that up automatically.

**Reply by hugary1995** [reply]
Oh and also update the conversion table in `Conversion.C::initQuadratureType()`.

**Reply by roystgnr** [reply]
On the libMesh side there's a few places to edit besides the new subclass's files and `enum_quadrature_type.h` - we'd need a case in `quadrature_build.C`, a map entry in `string_to_enum.C`, and we'd really want an instantiation in `quadrature_test.C` except that I'm not sure the tests there are appropriate for something that's underintegrating in that way.

Off topic from the conversation here, but I don't like that manual code in `Conversion.C`.  It has to be edited to add new enums now, but ideally we wouldn't be having to change anything at all in MOOSE.  C++ doesn't even support enough reflection to iterate over a sparse `enum`'s valid values, but maybe libMesh should provide some kind of container (a `constexpr initializer_list`??) of enum values and then `Conversion.C` could iterate over that and do `enum_to_string` (and prepend that extra `Q`) with each.  We have to write string conversions in libMesh for new enum entries, but we ought to be able to get away with not writing a redundant isomorphic conversion elsewhere.

**Reply by hugary1995** [reply]
> Off topic from the conversation here, but I don't like that manual code in Conversion.C. It has to be edited to add new enums now, but ideally we wouldn't be having to change anything at all in MOOSE. C++ doesn't even support enough reflection to iterate over a sparse enum's valid values, but maybe libMesh should provide some kind of container (a constexpr initializer_list??) of enum values and then Conversion.C could iterate over that and do enum_to_string (and prepend that extra Q) with each. We have to write string conversions in libMesh for new enum entries, but we ought to be able to get away with not writing a redundant isomorphic conversion elsewhere.

Yep, you are apparently on top of this https://github.com/idaholab/moose/issues/17810

**Reply by roystgnr** [reply]
For some definition of "on top of".  If I had a long-term memory I'd be dangerous.
